src/System.Management.Automation/engine/Utils.cs

```diff
@@ -33,6 +34,215 @@ namespace System.Management.Automation
+        /// </summary>
+        /// <param name="digits">Span or string of binary digits. Assumes all digits are either 1 or 0.</param>
+        /// <param name="unsigned">
+        /// Whether to treat the number as unsigned. When false, respects established conventions
+        /// with sign bits for certain input string lengths.
+        /// </param>
+        /// <returns>Returns the value of the binary string as a BigInteger.</returns>
+        internal static BigInteger ParseBinary(ReadOnlySpan<char> digits, bool unsigned)
+        {
+            if (!unsigned)
+            {
+                if (digits[0] == '0')
+                {
+                    unsigned = true;
+                }
+                else
+                {
+                    switch (digits.Length)
+                    {
+                        // Only accept sign bits at these lengths:
+                        case 8: // byte
+                        case 16: // short
+                        case 32: // int
+                        case 64: // long
+                        case 96: // decimal
+                        case int n when n >= 128: // BigInteger
+                            break;
+                        default:
+                            // If we do not flag these as unsigned, bigint assumes a sign bit for any (8 * n) string length
+                            unsigned = true;
+                            break;
+                    }
+                }
+            }
+
+            // Calculate number of 8-bit bytes needed to hold the input,  rounded up to next whole number.
+            // This value will also be used as the indexer for our array.
+            int outputByteWalker = (digits.Length + 7) / 8;
+            Span<byte> outputBytes = new byte[outputByteWalker--];
+
+            // We need to be prepared for any partial leading bytes, (e.g., 010|00000011|00000100|00000101)
+            // or for lengths < 8 (less than one byte) e.g., 010.
+            // Walk right to left, stepping one whole byte at a time (if there are any whole bytes).
+            int blockWalker;
+            for (blockWalker = digits.Length - 1; blockWalker >= 7; blockWalker -= 8)
+            {
+                // We use bit shifts and binary-or to sum the values in each byte.  These calculations will actually
+                // create values higher than a single byte, but the higher bits are quietly stripped out when cast
+                // to byte.
+                // '1' == 49 (00110001); '0' == 48 (00110000); we can safely bitshift the whole thing and then
+                // simply discard the unneeded higher bits with the cast. The low bits are added in separately to
+                // allow us to strip the higher 'noise' bits before we sum the values using binary-or.
+                outputBytes[outputByteWalker--] =
+                    (byte)(((digits[blockWalker - 7] << 7)
```

Comment:
We could consider Span and Slice by 8 here. Then in a temp method we could pack the shift operations in cycle (like in line 235 - there we could use the temp method too). 
We get definitely better readability and I'd expect better performance.

Reply:
Would different formatting help with readability, or is it because of the bitwise operations themselves? This calculation is not as readable as a loop, but it is *very* fast.

Using a loop, like in line 235, and dotNetBench testing:

- a random input of up to 32 characters, this calculation comes out at 64ns, a loop at 80ns. 
- a random input of up to 192 characters, this calculation comes out at ~250ns, the loop version at ~330ns.

(but if you can get a loop to go faster, that would be even better).

We did format it for better readability, but StyleCop did not like the extra spaces. It used to look like this, showing two separate groups of operations being combined:

```
                outputBytes[outputByteWalker--] = (byte)
                   (((digits[blockWalker - 7] << 7) |
                     (digits[blockWalker - 6] << 6) |
                     (digits[blockWalker - 5] << 5) |
                     (digits[blockWalker - 4] << 4))
                 |
                   (((digits[blockWalker - 3] << 3) |
                     (digits[blockWalker - 2] << 2) |
                     (digits[blockWalker - 1] << 1) |
                      digits[blockWalker    ]      ) & 0b1111)
                );   
```

Reply:
I agree with current code. My request is more to do measurements like you have done. I am thinking more about readability. Perhaps cycle is more slow because of having more of jumping. Perhaps we could speed up the cycle by skipping '0' digits and using cache table for getting masks.

Reply:
I am not entirely sure how we would go about coding that... But I'm all for eking out the best performance possible with better readability!

Reply:
1. In current code we shift zero - we could skip zero digits because this doesn't change result byte.
2. In current code we shift one - we could get prepared values from a cache.

```c#
for (int i = 0; i <= 7; i++) 
{ 
    if (digits[i] == '1')
    {
       currentByteValue |= cache[i]; 
    }
}
```

Reply:
We do still need to shift left by one when we hit a zero digit; `0b10` is `2`, whereas `0b1` is `1` -- a zero bit still indicates a doubling of value.

For those, there is no `|=` operation that would work. Using a cache sounds like a good possibility also, but given that there are numerals that may be handled here that are only possible as `BigInteger` (e.g., anything above `2**96` or thereabouts) we'd possibly need to branch here? That might be enough to make it not worth it, but we'd have to test it to be sure.

Reply:
Thanks for the experiment. Cycle unroll still works!
I suggest to add comment about this in the code to avoid changing it late.

Reply:
So apparently it can parse binary with 0 allocation. I've tested the code and it seems to *work*, I just have no idea how that's supposed to work, heh. Very spiffy.

But no, the loops are consistently *much* worse off in every way, unfortunately.

Reply:
@iSazonov 

@HumanEquivalentUnit and I ran a couple more tests after rerolling the parser into a method using `.Slice()` to traverse the input rather than by indexing -- just for thoroughness, really.

(Results](https://gist.github.com/vexx32/0b145d34a62ea2c8371da51baba352fe) - [Code for new method](https://gist.github.com/HumanEquivalentUnit/f17e76ffd2109a306fb6a673a6cba33f)

Reply:
Thanks! I think you could get additional win with:
1. Moving digits.Lenght to local var
2. digits = digits.Slice(0, digits.Length - 8); -> digits = digits.Slice(0, digits.Length - 9);

Reply:
Slice is `Span.Slice(int startIndex, int length)` so `- 9` would skip bits. 

Reran the test with a local var for length, and it's more consistent in the processing time, but still notably slower than the current method (otherwise roughly the same, really).